[CS] Allow ExprPatterns to be type-checked in the solver

hamishknight opened the pull request:
Previously we would wait until CSApply, which would trigger their type-checking in `coercePatternToType`. This caused a number of bugs, and hampered solver-based completion, which does not run CSApply. Instead, type-check ExprPatterns as a part of solving the overall pattern. We can then modify `coercePatternToType` to accept a closure, which allows the solver to take over rewriting the ExprPatterns it has already solved.

This then sets the stage for the complete removal of `coercePatternToType`, and doing all pattern type-checking in the solver.

This PR also includes a couple of cleanups/fixes for `getTypeForPattern`, and some other pattern crasher fixes.

Resolves #61850
Resolves #63374
rdar://104954155
rdar://105782480
rdar://106598067
rdar://107709341
rdar://107420031
rdar://109419240

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented on test/Constraints/patterns.swift:
```diff
@@ -564,3 +564,84 @@ struct TestIUOMatchOp {
+  if case { (); return x }() = x {}
+}
+
+func testExprPatternIsolation() {
```
@xedin Mind double checking that these cases make sense?

hamishknight commented:
@swift-ci please test

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4178,7 +4178,6 @@ GenericFunctionType *GenericFunctionType::get(GenericSignature sig,
                                               Type result,
                                               Optional<ExtInfo> info) {
   assert(sig && "no generic signature for generic function type?!");
-  assert(!result->hasTypeVariable());
```
As long as we allocate the GenericFunctionType in the correct arena it is probably fine. However we cannot allow type variables in the GenericSignature for more fundamental reasons.

xedin commented on test/Constraints/patterns.swift:
```diff
@@ -564,3 +564,84 @@ struct TestIUOMatchOp {
+  // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
+  // expected-note@-3 {{requirement from conditional conformance of 'Any?' to 'Equatable'}}
+
+  // Though we will try Double for an integer literal...
```
Is this specific to `Double` or `Int` works as well (I expect it does)?

xedin reviewed (approved):
LGTM! It might make sense to have an alternative version of this where we either use a TypeJoinExpr to type-check expr patterns that convert to the same generic parameter or simply enable bi-directional checking for patterns since each pattern is solved separately anyway and it doesn't seem like they are too complex.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/Sema/TypeCheckPattern.cpp:
```diff
@@ -1034,15 +1034,61 @@ void repairTupleOrAssociatedValuePatternIfApplicable(
+      EEP->setType(patternTy);
+      return EEP;
+    } else {
+      // ...but for non-optional types it can never match! Diagnose it.
```
Yes, but we also need to diagnose for cases that don't yet go through the constraint system. We should be able to move this into the constraint system once we delete `coercePatternToType`

xedin replied:
Fair enough! We could start with cases that do go through the solver and expand later too :)

hamishknight commented:
Source compat failures match those on main

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight committed:
[CS] Fix a couple of constraints in `getTypeForPattern`

The TypedPattern and IsPattern constraints were
incorrectly written, with conversions propagating
out of the patterns, when really conversions
ought to propagate into patterns. In any case, it
seems like we really want equality here. Fix the
constraints to use equality, and have the cast
constraint operate on the external pattern type
instead of the subpattern type.

hamishknight committed:
[CS] Reverse the type order in a couple of pattern equality constraints

Order them such that if they were changed to
conversions, they would be sound. This shouldn't
make a difference, but unfortunately it turns out
pattern equality is not symmetric. As such, tweak
the pattern equality logic to account for the
reversed types. This allows us to remove a special
case from function matching.

hamishknight committed:
[CS] Remove external type logic from `getTypeForPattern`

This shouldn't be necessary, we should be able to
solve with type variables instead. This makes sure
we don't end up with weird special cases that only
occur when an external type is present.

hamishknight committed:
[CS] Remove null pattern handling from `getTypeForPattern`

Push the only null case that can occur up into the
caller.

hamishknight committed:
[Sema] Walk SyntacticElementTarget for completion

Instead of walking the single ASTNode from the
target, walk all AST nodes associated with the
target to find the completion expr. This is needed
to find the completion expr in a pattern for an
initialization target.

hamishknight committed:
[CS] Remove null Type handling from `getTypeForPattern`

We should never CSGen a null Type for patterns.

hamishknight committed:
[CS] Don't form conversion between switch subject and pattern

This is wrong because there's nowhere to put any
conversion that is introduced, meaning that we'll
likely crash in SILGen. Change the constraint to
equality, which matches what we do outside of the
constraint system.

rdar://107709341

hamishknight committed:
[CS] Allow ExprPatterns to be type-checked in the solver

Previously we would wait until CSApply, which
would trigger their type-checking in
`coercePatternToType`. This caused a number of
bugs, and hampered solver-based completion, which
does not run CSApply. Instead, form a conjunction
of all the ExprPatterns present, which preserves
some of the previous isolation behavior (though
does not provide complete isolation).

We can then modify `coercePatternToType` to accept
a closure, which allows the solver to take over
rewriting the ExprPatterns it has already solved.

This then sets the stage for the complete removal
of `coercePatternToType`, and doing all pattern
type-checking in the solver.

hamishknight committed:
[CS] Improve diagnostics a bit for pattern mismatch

There's still plenty of more work to do here for
pattern diagnostics, including introducing a
bunch of new locator elements, and handling things
like argument list mismatches. This at least lets
us fall back to a generic mismatch diagnostic.

hamishknight committed:
[CS] Fix `coercePatternToType` enum cast handling

Previously if the cast was unresolved, we would
emit a warning and bail with `nullptr`. This is
wrong, because the caller expects a `nullptr`
return to mean we emitted an error. Change the
diagnostic to an error to fix this. This may
appear source breaking, but in reality previously
we were failing to add the cast at all in this case,
which lead to a crash in SILGen. We really do
want to reject these cases as errors, as this
will give us a better opportunity to fall back to
type-checking as ExprPatterns, and better matches
the constraint solver type-checking.

Also while we're here, change the diagnostic for
the case where we don't have an existential context
type from the confusing "enum doesn't have member"
diagnostic to the pattern mismatch diagnostic.

rdar://107420031

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@xedin Mind taking a look at 406354935e93ebff5a9c08b7b9055e3560ee77e6? I _believe_ that should be the only change since you last looked at this

hamishknight commented:
@swift-ci please smoke test and merge

hamishknight commented:
Going to stick to the conjunction implementation for now

hamishknight commented:
@swift-ci please smoke test and merge

hamishknight committed:
[test] Add test case that exercises lots of `~=` overloads

hamishknight commented:
@swift-ci please smoke test and merge
